licensing/src/main/java/io/crate/license/LicenseService.java

```diff
@@ -0,0 +1,248 @@
+                        }
+
+                        @Override
+                        public void onFailure(Exception e) {
+                            logger.error("Unable to register license", e);
+                        }
+                    });
+            }
+        }
+    }
+
+    private static byte[] decryptLicenseContent(LicenseType type, byte[] encryptedContent) {
+        byte[] decryptedContent = null;
+        switch (type) {
+            case TRIAL:
+                decryptedContent = TrialLicense.decrypt(encryptedContent);
+                break;
+            case ENTERPRISE:
+                decryptedContent = EnterpriseLicense.decrypt(encryptedContent);
+                break;
+            default:
+        }
+        return decryptedContent;
+    }
+
+    @Override
+    protected void doStop() {
+    }
+
+    @Override
+    protected void doClose() {
+        clusterService.removeListener(this);
+        currentLicense.set(null);
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        ClusterState currentState = event.state();
+
+        if (currentState.blocks().hasGlobalBlock(GatewayService.STATE_NOT_RECOVERED_BLOCK)) {
+            return;
+        }
+
+        LicenseKey previousLicenseKey = getLicenseMetadata(event.previousState());
+        LicenseKey newLicenseKey = getLicenseMetadata(currentState);
+
+        if (previousLicenseKey == null && newLicenseKey == null) {
+            registerTrialLicense(currentState);
+            return;
+        }
+
+        if (newLicenseKey != null && !newLicenseKey.equals(previousLicenseKey)) {
+            try {
+                DecryptedLicenseData decryptedLicenseData = licenseData(decodeLicense(newLicenseKey));
+                LicenseExpiryNotification expiryNotification = getLicenseExpiryNotification(decryptedLicenseData);
+
+                if (expiryNotification != null) {
+                    long millisToExpiration = decryptedLicenseData.millisToExpiration();
+                    if (expiryNotification.equals(LicenseExpiryNotification.EXPIRED)) {
+                        logger.error(expiryNotification.notificationMessage(millisToExpiration));
+                    } else if (expiryNotification.equals(LicenseExpiryNotification.SEVERE)) {
+                        logger.error(expiryNotification.notificationMessage(millisToExpiration));
+                    } else if (expiryNotification.equals(LicenseExpiryNotification.MODERATE)) {
+                        logger.warn(expiryNotification.notificationMessage(millisToExpiration));
+                    }
+                }
+
+                this.currentLicense.set(decryptedLicenseData);
+            } catch (IOException e) {
+                logger.error("Received invalid license. Unable to read the license data.");
+            }
+        }
+    }
+
+    @Override
+    public void onSuccess(ClusterState build) {
```

Comment:
Are these coming from `GatewayStateRecoveredListener ` ? Could that be removed?

Reply:
Ah yes, great catch!